diff --git a/configure b/configure
index 80ede83..19d6e1f 100755
--- a/configure
+++ b/configure
@@ -769,40 +769,27 @@ if test -z "$target_list" ; then
     exit 1
 fi
 
-if test -z "$cross_prefix" ; then
-
 # ---
 # big/little endian test
 cat > $TMPC << EOF
-#include <inttypes.h>
-int main(int argc, char ** argv){
-        volatile uint32_t i=0x01234567;
-        return (*((uint8_t*)(&i))) == 0x67;
+short big_endian[] = { 0x4269, 0x4765, 0x4e64, 0x4961, 0x4e00, 0, };
+short little_endian[] = { 0x694c, 0x7454, 0x654c, 0x6e45, 0x6944, 0x6e41, 0, };
+extern int foo(short *, short *);
+int main(int argc, char *argv[]) {
+    return foo(big_endian, little_endian);
 }
 EOF
 
-if $cc $ARCH_CFLAGS -o $TMPE $TMPC > /dev/null 2> /dev/null ; then
-$TMPE && bigendian="yes"
-else
-echo big/little test failed
-fi
-
+if $cc $ARCH_CFLAGS -c -o $TMPO ${OS_CFLAGS} $TMPC > /dev/null 2>&1 ; then
+    if grep -q BiGeNdIaN $TMPO ; then
+        bigendian="yes"
+    elif grep -q LiTtLeEnDiAn $TMPO ; then
+        bigendian="no"
+    else
+        echo big/little test failed
+    fi
 else
-
-# if cross compiling, cannot launch a program, so make a static guess
-if test "$cpu" = "armv4b" \
-     -o "$cpu" = "hppa" \
-     -o "$cpu" = "m68k" \
-     -o "$cpu" = "mips" \
-     -o "$cpu" = "mips64" \
-     -o "$cpu" = "ppc" \
-     -o "$cpu" = "ppc64" \
-     -o "$cpu" = "s390" \
-     -o "$cpu" = "sparc" \
-     -o "$cpu" = "sparc64"; then
-    bigendian="yes"
-fi
-
+    echo big/little test failed
 fi
 
 # host long bits test
@@ -913,6 +900,23 @@ else
     fi
 fi # -z $sdl
 
+sdl_win32_libs=""
+if test "$mingw32" = "yes" -a "$sdl_static" = "yes"; then
+cat > $TMPC <<EOF
+#include <windows.h>
+#include <mmsystem.h>
+#include <wingdi.h>
+int main(void) { SwapBuffers(NULL);return (int)timeGetTime(); }
+EOF
+    if $cc $ARCH_CFLAGS -o $TMPE ${OS_CFLAGS} $TMPC $LDFLAGS -lwinmm -lgdi32 > /dev/null 2>&1 ; then
+       sdl_win32_libs="-lwinmm -lgdi32"
+    else
+       sdl=no
+       sdl_static=no
+       echo "SDL static linking not supported"
+    fi
+fi
+
 if test "$sdl" = "yes" ; then
 cat > $TMPC <<EOF
 #include <SDL.h>
@@ -1197,12 +1201,29 @@ if test "$pthread" = yes; then
   pthread=no
 cat > $TMPC << EOF
 #include <pthread.h>
-int main(void) { pthread_create(0,0,0,0); return 0; }
+static void *f(void *p) { return NULL; }
+int main(void) {
+  pthread_t thread;
+  pthread_create(&thread, 0, f, 0);
+  return 0;
+}
 EOF
+  pthread_static=""
+  pthread_flags=""
+  pthread_libs=""
+  if test "$static" = "yes"; then
+    pthread_static="-static"
+    if test "$mingw32" = "yes" ; then
+      pthread_flags="-DPTW32_STATIC_LIB"
+      pthread_libs="-lwsock32"
+    fi
+  fi
   for pthread_lib in $PTHREADLIBS_LIST; do
-    if $cc $ARCH_CFLAGS -o $TMPE $TMPC $pthread_lib 2> /dev/null > /dev/null ; then
+    if $cc $ARCH_CFLAGS -o $TMPE $pthread_flags $TMPC $pthread_static $pthread_lib $pthread_libs 2> /dev/null > /dev/null ; then
       pthread=yes
+      CFLAGS="$CFLAGS $pthread_flags"
       PTHREADLIBS="$pthread_lib"
+      LDFLAGS="$LDFLAGS $pthread_libs"
       break
     fi
   done
@@ -1442,13 +1463,17 @@ if test "$pcap" = "yes" ; then
   pcap=no
   cat > $TMPC << EOF
 #include <pcap.h>
-int main(void) { return pcap_lib_version(); }
+int main(void) { return (int)pcap_lib_version(); }
 EOF
-  if $cc $ARCH_CFLAGS -o $TMPE $PCAPLIBS $TMPC 2> /dev/null; then
+  if $cc $ARCH_CFLAGS -o $TMPE $TMPC $PCAPLIBS 2> /dev/null; then
   pcap=yes
   fi
 fi
 
+if test "$pcap" = "no" ; then
+  PCAPLIBS=""
+fi
+
 echo "Install prefix    $prefix"
 echo "BIOS directory    $prefix$datasuffix"
 echo "binary directory  $prefix$binsuffix"
@@ -1506,6 +1531,7 @@ echo "Documentation     $build_docs"
 echo "uname -r          $uname_release"
 echo "NPTL support      $nptl"
 echo "vde support       $vde"
+echo "pthread           $pthread"
 echo "AIO support       $aio"
 echo "IO thread         $io_thread"
 echo "Install blobs     $blobs"
@@ -1719,11 +1745,11 @@ if test "$sdl1" = "yes" ; then
   echo "#define CONFIG_SDL 1" >> $config_host_h
   echo "CONFIG_SDL=y" >> $config_host_mak
   if test "$target_softmmu" = "no" -o "$static" = "yes"; then
-    echo "SDL_LIBS=$sdl_static_libs" >> $config_host_mak
+    echo "SDL_LIBS=$sdl_static_libs $sdl_win32_libs" >> $config_host_mak
   elif test "$sdl_x11" = "yes" ; then
-    echo "SDL_LIBS=`$sdl_config --libs` -lX11" >> $config_host_mak
+    echo "SDL_LIBS=`$sdl_config --libs` -lX11 $sdl_win32_libs" >> $config_host_mak
   else
-    echo "SDL_LIBS=`$sdl_config --libs`" >> $config_host_mak
+    echo "SDL_LIBS=`$sdl_config --libs` $sdl_win32_libs" >> $config_host_mak
   fi
   if [ "${aa}" = "yes" ] ; then
     echo "SDL_CFLAGS=`$sdl_config --cflags` `aalib-config --cflags`" >> $config_host_mak
